Styhead/rz-cmn: Enhancements and Workarounds for RZG2L DSI

long-trinh-rvc opened the pull request:
Changes include:
- Adding atomic_pre_enable support in RZ-DU.
- Implementing MIPI DSI host transfer support.
- Early sequence transfer workaround for better initialization.

long-trinh-rvc committed:
drm: renesas: rz-du: Add atomic_pre_enable

When drm_panel.prepare_prev_first is set to true in a panel driver, the
panel expects the MIPI DSI hardware to be already configured before the
panel's prepare function is called because it might need to send DCS
commands.

Signed-off-by: Chris Brandt <chris.brandt@renesas.com>
Signed-off-by: Long Trinh <long.trinh.rh@bp.renesas.com>

long-trinh-rvc committed:
drm: renesas: rz-du: Implement MIPI DSI host transfers

Add support for sending MIPI DSI command packets from the host to a
peripheral. This is required for panels that need configuration before
they accept video data.

Also for long reads to work properly, set DCS maximum return packet size
to the value of the DMA buffer size.

Based on Renesas Linux kernel v5.10 repos [1].

Link: https://github.com/renesas-rz/rz_linux-cip.git
Cc: Biju Das <biju.das.jz@bp.renesas.com>
Signed-off-by: Hugo Villeneuve <hvilleneuve@dimonoff.com>
Signed-off-by: Long Trinh <long.trinh.rh@bp.renesas.com>

long-trinh-rvc committed:
drm: renesas: rz-du: Workaround to transfer sequence earlier

This update adds support for detecting whether the connected entity
is a panel or a bridge in the rzg2l_mipi_dsi driver.

If a panel is detected, the driver now performs an early sequence to
prepare and enable the panel before starting video transmission,
ensuring proper initialization and avoiding timing issues.

Signed-off-by: Long Trinh <long.trinh.rh@bp.renesas.com>

tiennguyenuh commented:
@long-trinh-rvc, This PR needs to be rebased again, please rebase onto the latest branch.

long-trinh-rvc commented:
> @long-trinh-rvc, This PR needs to be rebased again, please rebase onto the latest branch.

I have updated this

long-trinh-rvc committed:
drm: panel: Add DCS init with fallback to original method

Implemented DCS init sequence with fallback for compatibility

Signed-off-by: Long Trinh <long.trinh.rh@bp.renesas.com>

vudangRVC commented on drivers/gpu/drm/panel/panel-toshiba-tc358762.c:
```diff
@@ -96,11 +98,125 @@ struct tc358762 {
+	return 0;
+}
+
+static const struct panel_init_cmd panel_5_a_init[] = {
```
What is the difference between these commands and the sequences in 
```static int tc358762_dsi_init(struct tc358762 *p)```

```c
tc358762_gen_write_seq(dsi, 0x10, 0x02, 0x03, 0x00, 0x00, 0x00);//LANE
tc358762_gen_write_seq(dsi, 0x64, 0x01, 0x0c, 0x00, 0x00, 0x00);//D0S_CLRSIPOCOUNT
```

Do we need to keep both?

long-trinh-rvc replied:
We use `panel_5_a_init` for panels that utilize DCS commands, and `tc358762_dsi_init` for panels that use generic command sequences; both are necessary to support different panel types. Only one of them will be used, depending on the panel type.

vudangRVC commented on drivers/gpu/drm/panel/panel-toshiba-tc358762.c:
```diff
@@ -526,6 +717,7 @@ static const struct drm_display_mode tc358762_mode = {
 
 static const struct bridge_desc tc358762_bridge = {
 	.desc = {
+		.init = NULL, // if panel support DCS command, replace by struct panel_init_cmd array
```
@long-trinh-rvc , I don't think directly modifying the driver source code just to switch between versions is a proper solution. I would recommend using the device tree for better flexibility. In this case, you can add an additional field to the device tree. For example, "panel-v2".

```c
mipi_panel: panel@0 {
                compatible = "rzsbc,tc358762";
                reg = <0>;
                panel-v2;
                status = "disabled";

                port {
                        panel_in: endpoint {
                                remote-endpoint = <&dsi0_out>;
                        };
                };
        };
```

long-trinh-rvc replied:
I have updated this in the latest commit. For example, we can set panel v2 in the DTS as below::

```
	mipi_panel: panel@0 {
		compatible = "rzsbc,tc358762";
		reg = <0>;
		status = "disabled";
		panel-id = <PANEL_5_V2>; 

		port {
			panel_in: endpoint {
				remote-endpoint = <&dsi0_out>;
			};
		};
	};
```

long-trinh-rvc committed:
drm: panel: tc358762: Add panel-id property to select init sequence

Introduce 'panel-id' DT property to allow runtime selection of
panel initialization commands.
This change moves the init sequence pointer from the static
panel_desc to a runtime field in struct tc358762, enabling
different panel versions (e.g. PANEL_5_V2) to be initialized
without modifying the descriptor.
Fallback to default behavior is preserved when 'panel-id' is
not specified.

Signed-off-by: Long Trinh <long.trinh.rh@bp.renesas.com>

vudangRVC commented on arch/arm64/boot/dts/renesas/rzg2l-sbc.dts:
```diff
@@ -869,6 +870,8 @@
 		compatible = "rzsbc,tc358762";
 		reg = <0>;
 		status = "disabled";
+		/* Example: to change the panel version, set panel-id = <PANEL_5_V1>; (default value) */
```
The documentation and guidelines should be described in the device tree documentation files. Please refer to other existing device tree documentation files for reference. There's no need to explain it here.
Since we don’t have a device tree documentation file for the panel yet, let’s remove these lines.

Also, please set the new field with a default value by default

long-trinh-rvc replied:
I've updated this

vudangRVC commented on drivers/gpu/drm/panel/panel-toshiba-tc358762.c:
```diff
@@ -595,6 +790,19 @@ int tc358762_dsi_probe(struct mipi_dsi_device *dsi)
+		switch (panel_id) {
+		case PANEL_5_V2:
+			panel->init = panel_5_a_init;
+			dev_dbg(&dsi->dev, "panel init as PANEL_5_V2\n");
```
"panel init as PANEL_5_V2\n". Why dont you use panel_id instead of "PANEL_5_V2"?

long-trinh-rvc replied:
This case is used for PANEL_5_V2. If I use panel_id, the debug log only shows a number, which can make the log harder to understand

vudangRVC commented on drivers/gpu/drm/panel/panel-toshiba-tc358762.c:
```diff
@@ -280,6 +426,44 @@ static int tc358762_dsi_init(struct tc358762 *p)
 }
 
+// Init Panel using DCS	commands
+static int tc358762_panel_init(struct tc358762 *p)
```
This function is intented to support dcs init. The current name can cause confusion. Because when I read it, it sounds like a generic function.
Please use a proper name.

long-trinh-rvc replied:
I will use `tc358762_panel_dcs_init` instead of `tc358762_panel_init`

vudangRVC commented on drivers/gpu/drm/panel/panel-toshiba-tc358762.c:
```diff
@@ -308,10 +500,11 @@ static int tc358762_prepare(struct drm_panel *panel)
 static int tc358762_enable(struct drm_panel *panel)
 {
 	struct tc358762 *p = to_tc358762(panel);
+	int ret;
```
This variable is unused

long-trinh-rvc replied:
I've updated this

long-trinh-rvc committed:
drm: panel: tc358762: Rename panel_init and set panel-id in DTS

Rename tc358762_panel_init() to tc358762_panel_dcs_init() for clarity
and update the RZ/G2L-SBC device tree to set panel-id to PANEL_5_V1 as default.

Signed-off-by: Long Trinh <long.trinh.rh@bp.renesas.com>
